core/src/main/java/org/apache/iceberg/Snapshots.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Pair;
+
+public class Snapshots {
+  private Snapshots() {
+  }
+
+  public static MicroBatchBuilder from(Snapshot snapshot, FileIO io) {
```

Comment:
Because this is built from a single snapshot, batches can't cross snapshot boundaries. Is that a good thing or a bad thing?

It is nice to process all data from a snapshot before moving on. But, this could cause a large variation in batch size and the last batch in every snapshot will probably be small.

Reply:
@rdblue I'm not sure if I understand correctly, though here micro batch always belongs to one snapshot, in Streaming Reader implementation, it will process multiple micro batches in one streaming batch (controlled by size per batch). So the variation seems not a problem AFAIK.

Reply:
Okay, I understand. So if a microbatch is the last one in a snapshot, the reader can get another microbatch from the next snapshot (requesting a smaller size) and run the two as a combined batch. And we could have a configuration setting to to make sure batches align with snapshot boundaries. Sounds good to me!

Reply:
> Okay, I understand. So if a microbatch is the last one in a snapshot, the reader can get another microbatch from the next snapshot (requesting a smaller size) and run the two as a combined batch. And we could have a configuration setting to to make sure batches align with snapshot boundaries.

Yes, that's the current implementation :).